File: C++_Version/Hash_Table/500Keyboard_Row.cpp
//Given a List of words, return the words that can be typed using letters of alphabet on only one row's of American keyboard like the image below.
//
//
//American keyboard
//
//
//Example 1:
//Input: ["Hello", "Alaska", "Dad", "Peace"]
//	Output : ["Alaska", "Dad"]
//	Note :
//	You may use one character in the keyboard more than once.
//	You may assume the input string will only contain letters of alphabet.


#include <vector>
#include <iostream>
#include <algorithm>
#include <cstdlib>
#include <functional>
#include <string>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
using namespace std;


class Solution {
public:
	vector<string> findWords(vector<string>& words) {

		unordered_set<char> keyboard_top_row, keyboard_middle_row, keyboard_bottom_row;

		string keyboard_top = "QWERTYUIOPqwertyuiop", keyboard_middle = "ASDFGHJKLasdfghjkl", keyboard_bottom = "ZXCVBNMzxcvbnm";

		for (int i = 0; i < (int)keyboard_top.length(); ++i)
			keyboard_top_row.insert(keyboard_top[i]);
		for (int i = 0; i < (int)keyboard_middle.length(); ++i)
			keyboard_middle_row.insert(keyboard_middle[i]);
		for (int i = 0; i < (int)keyboard_bottom.length(); ++i)
			keyboard_bottom_row.insert(keyboard_bottom[i]);


		vector<string> output;
		int words_size = words.size();
		for (int i = 0; i < words_size; ++i)
		{
			bool flag = true;

			int length = words[i].length();

			if (length == 0 || length == 1)
				output.push_back(words[i]);
			else
			{
				if (keyboard_top_row.find(words[i][0]) != keyboard_top_row.end())
				{
					for (int j = 1; j < length; ++j)
					{
						if (keyboard_top_row.find(words[i][j]) == keyboard_top_row.end())
						{
							flag = false;
							break;
						}	
					}
				}
				else if (keyboard_middle_row.find(words[i][0]) != keyboard_middle_row.end())
				{
					for (int j = 1; j < length; ++j)
					{
						if (keyboard_middle_row.find(words[i][j]) == keyboard_middle_row.end())
						{
							flag = false;
							break;
						}
					}
				}
				else if (keyboard_bottom_row.find(words[i][0]) != keyboard_bottom_row.end())
				{
					for (int j = 1; j < length; ++j)
					{
						if (keyboard_bottom_row.find(words[i][j]) == keyboard_bottom_row.end())
						{
							flag = false;
							break;
						}
					}
				}

				if (flag)
					output.push_back(words[i]);
			}
		}

		return output;
	}
};

int main()
{
	vector<int> list1({ 1, 1, 2, 2, 3, 3 });
	vector<string> list2({ "Hello", "Alaska", "Dad", "Peace" });

	Solution sol1;

	vector<string> receive = sol1.findWords(list2);

	system("pause");

	return 0;
}

File: Java_Version/500Keyboard_Row.java
/*Given a List of words, return the words that can be typed using letters of alphabet on only one row's of American keyboard like the image below.


        American keyboard


        Example 1:
        Input: ["Hello", "Alaska", "Dad", "Peace"]
        Output: ["Alaska", "Dad"]
        Note:
        You may use one character in the keyboard more than once.
        You may assume the input string will only contain letters of alphabet.*/


import java.util.*;
import java.lang.Math;
import java.lang.System;
import java.lang.Integer;


public class Main {

    public static void main(String[] args) throws ArithmeticException {

        int a = 3, b = 0;


        int result = a / b;


        System.out.println("haha");
    }

}


class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}


class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}


class Solution {

    private ArrayList<Integer> resultIndex;
    private HashMap<Character, Integer> keyboardTable = new HashMap<>();

    public String[] findWords(String[] words) {

        int wordsLength = words.length;

        resultIndex = new ArrayList<Integer>(wordsLength);
        initKeyboardTable();

        for (int i = 0; i < wordsLength; i++) {
            boolean flag = processString(words[i]);
            if (flag) {
                resultIndex.add(i);
            }
        }

        String[] result = new String[resultIndex.size()];
        for (int i = 0; i < result.length; i++){
            result[i] = words[resultIndex.get(i)];
        }

        return result;
    }

    private boolean processString(String word) {
        int value = keyboardTable.get(word.charAt(0));

        int wordLength = word.length();
        for ( int i = 1; i < wordLength; i++ ){
            if ( keyboardTable.get(word.charAt(i)) != value ){
                return false;
            }
        }

        return true;
    }

    private void initKeyboardTable() {
        keyboardTable.put('Q',1);
        keyboardTable.put('W',1);
        keyboardTable.put('E',1);
        keyboardTable.put('R',1);
        keyboardTable.put('T',1);
        keyboardTable.put('Y',1);
        keyboardTable.put('U',1);
        keyboardTable.put('I',1);
        keyboardTable.put('O',1);
        keyboardTable.put('P',1);
        keyboardTable.put('q',1);
        keyboardTable.put('w',1);
        keyboardTable.put('e',1);
        keyboardTable.put('r',1);
        keyboardTable.put('t',1);
        keyboardTable.put('y',1);
        keyboardTable.put('u',1);
        keyboardTable.put('i',1);
        keyboardTable.put('o',1);
        keyboardTable.put('p',1);

        keyboardTable.put('A',2);
        keyboardTable.put('S',2);
        keyboardTable.put('D',2);
        keyboardTable.put('F',2);
        keyboardTable.put('G',2);
        keyboardTable.put('H',2);
        keyboardTable.put('J',2);
        keyboardTable.put('K',2);
        keyboardTable.put('L',2);
        keyboardTable.put('a',2);
        keyboardTable.put('s',2);
        keyboardTable.put('d',2);
        keyboardTable.put('f',2);
        keyboardTable.put('g',2);
        keyboardTable.put('h',2);
        keyboardTable.put('j',2);
        keyboardTable.put('k',2);
        keyboardTable.put('l',2);

        keyboardTable.put('Z',3);
        keyboardTable.put('X',3);
        keyboardTable.put('C',3);
        keyboardTable.put('V',3);
        keyboardTable.put('B',3);
        keyboardTable.put('N',3);
        keyboardTable.put('M',3);
        keyboardTable.put('z',3);
        keyboardTable.put('x',3);
        keyboardTable.put('c',3);
        keyboardTable.put('v',3);
        keyboardTable.put('b',3);
        keyboardTable.put('n',3);
        keyboardTable.put('m',3);
    }
}